### PR TITLE
remove wasted allocation space for JL_ARRAY_IMPL_NUL

### PR DESCRIPTION
**Pull request by vtjnash**
Handle this more correctly in Cstring conversion instead. Although apparently this feature and test has not worked in Julia since String became a first class object (and not wrapping Array).

Refs: cee4cb78da8f5cfa34808e3310e2366769bc3245

### CODE REVIEW

File: base/c.jl [hunk 1]
```diff
@@ -212,6 +212,11 @@ containsnul(s::String) = containsnul(unsafe_convert(Ptr{Cchar}, s), sizeof(s))
 containsnul(s::AbstractString) = '\0' in s
 
 function unsafe_convert(::Type{Cstring}, s::Union{String,AbstractVector{UInt8}})
+    if !(s isa String)
+        # make sure s is null-terminated
+        push!(s, 0x00)
+        pop!(s)
```
Comment (barucden) [inline base/c.jl:218]: Just curious – how does this ensure that `s` is null-terminated? A `0x00` is pushed but popped immediately, so the content of `s` should be the same as at the beginning, no?

Comment (vtjnash) [reply base/c.jl:218]: A pop does not reverse the memory effects of a push, just the change to the length

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks(!"scalar", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51275#issuecomment-1716062355) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/fcb1fe5_vs_832e46d/report.md).

**Pull request merged**